In [25]:
import sys
sys.path.append("./../src/")

In [26]:
import random
seed = 0
random.seed(seed)

import pandas as pd
import numpy as np

from utilities import DATA_PATH, RESPONSE_COL_NAME

In [27]:
# Read in preprocessed data
df = pd.read_csv(f"{DATA_PATH}/data.csv")
display(df)

lapse_ape_grp_fe5fb8  stat_flag_LAPSED  is_Singapore  \
0                       0.0                 0             1   
1                       0.0                 0             1   
2                       0.0                 0             1   
3                       0.0                 0             1   
4                       0.0                 0             1   
...                     ...               ...           ...   
17979                   0.0                 1             1   
17980                   0.0                 0             1   
17981                   0.0                 0             1   
17982                   0.0                 0             1   
17983                   0.0                 0             1   

       lapse_ape_grp_e91421  recency_hlthclaim_unsuccess  \
0                       0.0                        False   
1                       0.0                        False   
2                       0.0                        False   
3                       0.0                        False   
4                       0.0                        False   
...                     ...                          ...   
17979                   0.0                        False   
17980                   0.0                        False   
17981                   0.0                        False   
17982                   0.0                        False   
17983                   0.0                        False   

       n_months_last_bought_32c74c  n_months_last_bought_grp_fe5fb8  \
0                           9999.0                           9999.0   
1                           9999.0                           9999.0   
2                           9999.0                           9999.0   
3                           9999.0                           9999.0   
4                           9999.0                           9999.0   
...                            ...                              ...   
17979                         36.0                           9999.0   
17980                       9999.0                           9999.0   
17981                       9999.0                           9999.0   
17982                       9999.0                           9999.0   
17983                       9999.0                           9999.0   

       flg_hlthclaim_14cb37_ever  race_desc_Indian  lapse_ape_lh_839f8a  ...  \
0                              1                 0                  0.0  ...   
1                              1                 0                  0.0  ...   
2                              1                 0                  0.0  ...   
3                              1                 0                  0.0  ...   
4                              1                 0                  0.0  ...   
...                          ...               ...                  ...  ...   
17979                          1                 0                  0.0  ...   
17980                          1                 0                  0.0  ...   
17981                          1                 0                  0.0  ...   
17982                          1                 0                  0.0  ...   
17983                          1                 1                  0.0  ...   

       n_months_last_bought_839f8a  f_mindef_mha  tot_inforce_pols  \
0                           9999.0             0               3.0   
1                           9999.0             0               1.0   
2                           9999.0             0               1.0   
3                           9999.0             0               1.0   
4                           9999.0             0               1.0   
...                            ...           ...               ...   
17979                       9999.0             0               1.0   
17980                       9999.0             0               1.0   
17981                       9999.0             0               1.0   
17982          

In [28]:
df[RESPONSE_COL_NAME].value_counts()

1    17280
0      704
Name: f_purchase_lh, dtype: int64

To split the data, we take 20% x 704 (140 samples) positive samples to the validation and test datasets each, as well as 140 negative samples. The rest go to the training dataset:

|          Train         |     Validation     |        Test        |
|------------------------|--------------------|--------------------|
|  424 positive samples  |140 positive samples|140 positive samples|
| 17000 negative samples |140 negative samples|140 negative samples|

In [29]:
n_train_positive, n_train_negative = 424, 17000
n_val_positive, n_val_negative = 140, 140
n_test_positive, n_test_negative = 140, 140

negative_indexes = np.where(df[RESPONSE_COL_NAME] == 1)[0].tolist()
positive_indexes = np.where(df[RESPONSE_COL_NAME] == 0)[0].tolist()

random.shuffle(negative_indexes)
random.shuffle(positive_indexes)

In [30]:
train_indices = positive_indexes[0 : n_train_positive] + negative_indexes[0 : n_train_negative]
val_indices = positive_indexes[n_train_positive : n_train_positive + n_val_positive] + negative_indexes[n_train_negative : n_train_negative + n_val_negative]
test_indices = positive_indexes[n_train_positive + n_val_positive : ] + negative_indexes[n_train_negative + n_val_negative : ]

train = df.iloc[train_indices, :].reset_index(drop=True)
val = df.iloc[val_indices, :].reset_index(drop=True)
test = df.iloc[test_indices, :].reset_index(drop=True)

In [31]:
train.to_csv(f"{DATA_PATH}/train.csv", index=False)
val.to_csv(f"{DATA_PATH}/val.csv", index=False)
test.to_csv(f"{DATA_PATH}/test.csv", index=False)